lib/puppet/util/rundeck_acl.rb

```diff
@@ -1,314 +1,235 @@
-      if !context.is_a? Hash
-        raise_err('context is not a Hash')
-      elsif context.empty?
-        raise_err('context is empty')
-      else
-        if context.keys.length != 1
-          raise_err('context can only contain project or application')
-        else
+        def validate_context(context)
+          raise_err('context is not a Hash') unless context.is_a? Hash
+          raise_err('context is empty') if context.empty?
+          raise_err('context can only contain project or application') unless context.keys.length == 1
           type = context.keys[0]
-          
-          case type
-          when 'project', 'application'
-            if !context[type].is_a? String
-              raise_err("context:#{type} is not a String")
-            end
-          else
-            raise_err("context can only be project or application")
-          end
-
+          raise_err("context:#{type} is not a String") unless context[type].is_a? String
+          raise_err('context can only be project or application') unless %w(application project).include? type
         end
-      end
-    end
-
-    def validate_rule_action(type, type_section, scope)
-      action_found = false
-      actions = []
-      property = ''
-      value = ''
 
-      if type_section.empty?
-        raise_err("for:#{type} is empty")
-      end
-      type_section.each do |e|
-        if !e.is_a? Hash
-          raise_err("for:#{type} entry is not a Hash")
-        end
-      end
+        def validate_rule_action(type, type_section, scope)
+          action_found = false
+          actions = []
+          property = ''
+          value = ''
 
-      type_section.each do |e|
-        e.each do |k,v|
-          if k.eql?('allow') or k.eql?('deny')
-            action_found = true
-            actions = v
-          elsif ['match','equals','contains'].include?(k)
-            case type
-            when 'resource'
-              property = v['kind']
-            when 'job'
-              if v['name']
-                property = 'name'
-                value = v['name']
-              elsif v['group']
-                property = 'group'
-                value = v['group']
-              else
-                property = v.keys[0]
-              end
-            when 'project'
-              if v['name']
-                property = 'name'
-                value = v['name']
-              else
-                property = v.keys[0]
-              end
-            when 'storage'
-              if v['name']
-                property = 'name'
-                value = v['name']
-              elsif v['path']
-                property = 'path'
-                value = v['path']
-              else
-                property = v.keys[0]
+          raise_err("for:#{type} is empty") if type_section.empty?
+          type_section.each do |e|
+            raise_err("for:#{type} entry is not a Hash") unless e.is_a? Hash
+          end
+          type_section.each do |e|
```

Comment:
this is some serious ugly code, and i'll have to take a look at this with more time and concentration…
it looks entirely too complex for its own good, and should probably be broken up in a couple of sub-functions